### DISCUSSION TITLE
Single Block Editor

### DISCUSSION THREAD

**Discussion by skttl** [umbraco/Umbraco-CMS#12927] [Features and ideas]
With [Nested Content being killed with the new backoffice](https://github.com/umbraco/rfcs/blob/main/cms/0024-implement-the-new-backoffice.md#property-editors-and-data-types), I really think we need a replacement for the [Single Item Mode](https://our.umbraco.com/Documentation/Fundamentals/Backoffice/property-editors/built-in-property-editors/Nested-Content/#single-item-mode) using blocks.

Today, I can create a block list, set a single possible block type, and set the min and max items to 1 to get somewhat close. But the editor still has to add the block themselves, and they can add more (just not publish). That's not what I want. I want a single embedded doctype in my doctype.

When I do the block list single item mode, I get this UI:
![image](https://user-images.githubusercontent.com/3726467/187617095-5ccaa6e4-fab4-48b7-a373-dd86c68ca9b3.png)

But I just want it to be like this:
![image](https://user-images.githubusercontent.com/3726467/187618001-90b560d4-b759-4f64-96d0-56b3bf114ca9.png)

The output type could be `BlockListItem<TContent, TSettings>` or maybe just `TContent` if no settings type is added.

**Comment by skttl** [comment]
Simple solution could be to add settings for the following
- create initial block automatically
- disable adding more blocks
- force open the block in inline editing mode - or, use a custom view for the block where the user can edit the content.

**Reply by drpeck** [reply]
Absolutely! I recall this being raised before. There was an edge case identified about why this is not a no-brainer. I wish I could remember what it was.

**Comment by FransdeJong** [comment]
Yes to this request.
Since we practically replaced nested content with blocklist it feels like a downgrade to not be able to use the single option.
This request removes an extra step for editors that feels unnecessary and also unclear why they have to do it.

**Comment by garpunkal** [comment]
I was looking on how to do yesterday! Perfect timing. 

We definitely need a single block instance, where it fulfills the points above.

**Comment by marcemarc** [comment]
I have literally hit this issue today, its a great technique for varying configuration for a page instead of having a ton of checkboxes that are on every page and get rarely used... They can all be replaced by a single option nested content item that can choose a single configuration type from different named configuration Element Types (or none at all...) if the configuration type then needs values they can be entered on the selected element type... Anyway added it today on a V10 site with Block List Editor in Inline Mode... It does not quite feel the same as nested content.. Mainly that you can add new blocks... Even though its configured to allow only one item... It would be super cool if that option disappeared when one block had been selected and the lock list was limited to one item

**Comment by Bakersbakebread** [comment]
Whilst not embedding, I put a PR in that at least stops them from doing something they're not allowed to do anyway! https://github.com/umbraco/Umbraco-CMS/pull/13056

**Comment by skttl** [comment]
I would love if we could have something like this in core, and I can work on a PR. Just need some pointers from HQ (@nul800sebastiaan @nielslyngsoe @filipbech )

- Would you accept a PR adding this?
- Would you prefer a new propert editor or additional settings on the block list as mentioned earlier
- If new datatype, how do you think the data should be saved? Could simply just be the same model as the Block List, just with one item in it - this would make it easy to switch between block list and "single block"

**Comment by nielslyngsoe** [comment]
Hi @skttl 

Thanks for highlighting this idea/feature.

I love this concept, actually it was also part of some initial material on Block Editors. So I think it will be good to have such.

I would encourage to choose between your inline-example and the option to edit in an overlay(infinite editor). For those who have a ton of options that can be hidden away.

I have some concerns on how this will work/look with features like:
* Reusable Blocks (Shared Blocks from DB, especially how to pick such? I guess that would be how to replace with such?)
* Copy/Paste, some might think of this as a Block feature, so we need to make that available somehow.

I have set up a meeting with Lasse(PO) for next week where we will dedicate some time to look at solutions and more importantly whether these should be requirements for a Block Editor that is in Backoffice.

If you already have some thoughts on how to enable these points, please pitch in.

Thanks :-)

**Comment by nielslyngsoe** [comment]
Hi @skttl and everyone reading along.

We had a talk today, Søren, Lasse and I.
We concluded that it would be easiest to extend the functionality of the Block List Editor.

We would like to provide a "Single mode" toggle in the PreValue Editor.
This PreValue editor should have a description describing that the output will change. (similar to what we do in Media Picker v.3)   

“Single Mode toggle“ Prevalue editor should be made as a custom PreValue editor. Cause we want it to validate the number of Blocks to make sure only a single Block is set when in this mode.

In the UI we will need to adjust some things when Single Mode is turned on.
The create buttons and remove buttons should go away.  The editor should automatically initialize the available Block. 

We also want to add a Property Action with a label of something like “Replace with an item from Clipboard”, as the Paste from Clipboard feature will be gone when we remove the Create buttons.


I hope this description makes sense, And I'm looking forward to seeing Sørens take on this.

Thanks in advance
